slash_commands/adorasac.js

```diff
@@ -0,0 +1,150 @@
+const { SlashCommandBuilder } = require('discord.js');
+const { isTower, allTowers, allUpgradeCrosspathSets, costOfTowerUpgradeSet, isTowerUpgrade } = require('../helpers/towers.js');
+const { red, cyber } = require('../jsons/colors.json');
+
+const XP_CAP = 400000;
+const XP_CAP_ADJUSTED = Math.ceil(XP_CAP / (4*5));
+
+// deprioritize water towers, illegal towers in CHIMPS, towers that are weird to sac in 2mpc, etc.
+const TOWER_PRIORITIES = new Map();
+TOWER_PRIORITIES.set('banana_farm', -1000); // illegal in chimps
+TOWER_PRIORITIES.set('monkey_sub', -700); // water tower
+TOWER_PRIORITIES.set('monkey_buccaneer', -850); // water tower
+TOWER_PRIORITIES.set('beast_handler', -620); // water tower in some paths
+TOWER_PRIORITIES.set('heli_pilot', -500); // can't be placed on midnight mansion
+TOWER_PRIORITIES.set('monkey_ace', -380); // weird in 2mp
+TOWER_PRIORITIES.set('sniper_monkey', -250); // weird in 2mp
+TOWER_PRIORITIES.set('druid_monkey', -120); // jungle druid is weird in 2mp
+
+builder = new SlashCommandBuilder()
+    .setName('adora-sac')
+    .setDescription('Calculate the optimal ewwww adora sacrifice to advance a certain amount of XP.')
+    .addIntegerOption((option) => option.setName('xp').setDescription('target XP to gain').setRequired(true).setMinValue(0).setMaxValue(XP_CAP))
+    .addStringOption((option) => option.setName('excluded_towers').setDescription('Comma-separated list of towers to exclude'));
+
+/**
+ * @param {string} excludedTowers raw excluded towers argument from command
+ * @returns {Set<string>} excluded tower upgrade ssets
+ */
+function parseExcludedTowers(excludedTowers) {
+    if (!excludedTowers) {
+        return new Set();
+    }
+
+    let result = new Set();
+    let excludedTowersList = excludedTowers.split(/\s*,\s*/);
+    const allUpgradeSets = allUpgradeCrosspathSets();
+
+    for (let excludedTower of excludedTowersList) {
+        let canonicalForm = Aliases.canonicizeArg(excludedTower);
+        if (isTower(canonicalForm)) {
+            for (let upgradeSet of allUpgradeSets) {
+                result.add(`${canonicalForm}#${upgradeSet}`);
+            }
+        } else if (isTowerUpgrade(canonicalForm, true)) {
+            result.add(canonicalForm);
```

Comment:
One slightly tricky aspect of the bot: the canonical form of an upgrade exists specifically with no crosspaths. That leads `excludedTowers.has(upgradeSetStr)` to return false below because `spike_factory#041 != spike_factory#040` in the following example:
![image](https://github.com/hemisemidemipresent/cyberquincy/assets/68867396/73d24de7-c59a-41b4-83f3-1f14c407fdb2)
I'm not totally sure how you want to handle this or if the current state is ok.

Reply:
If you ignore crosspath below, then specifically excluding `spike_storm#041` will also exclude `spike_storm#040` which is also imperfect. However, I guess at this point, these might be nitpicks..

Reply:
Ig the solution I am thinking of is to exclude upgrades building off of the specified excluded upgrade (as if you don't want a specific upgrade set you don't want upgrades building off of that either); for example, if I exclude `dart#022` then I also exclude `dart#032`, `dart#024`, etc. but not `dart#005` or `dart#012`.

Reply:
Got it, that's a good solution!